lib/puppet/network/http/connection.rb

```diff
@@ -189,6 +190,36 @@ def request_with_redirects(request, options)
             request.each do |header, value|
               current_request[header] = value
             end
+          when 429, 503
+            retry_after = current_response['Retry-After']
+
+            if retry_after.nil?
+              # Set return value and advance to next loop.
+              response = current_response
+              next
+            end
+
+            retry_sleep = parse_retry_after_header(retry_after)
+
+            if retry_sleep.nil?
+              Puppet.err(_('Received a %{status_code} response from the server, but the Retry-After header value of "%{retry_after}" could not be converted to an integer or RFC 2822 Date.') %
+                         {status_code: current_response.code,
+                          retry_after: retry_after.inspect})
+
+              response = current_response
+              next
+            end
+
+            if (retry_sleep > 0)
+              # Cap maximum sleep at the run interval of the Puppet agent.
+              retry_sleep = [retry_sleep, Puppet[:runinterval]].min
+
+              Puppet.warning(_('Received a %{status_code} response from the server. Sleeping for %{retry_sleep} seconds before retrying the request.') %
+                             {status_code: current_response.code,
+                              retry_sleep: retry_sleep})
+
+              ::Kernel.sleep(retry_sleep) if (retry_sleep > 0)
```

Comment:
This could happen if the server returned a RFC 2822 date that had already passed. Although, it might make more sense to have the `parse_retry_after_header` function return 0 in this case.